packages/engine-server/src/workspace/service.ts

```diff
@@ -859,6 +877,84 @@ export class WorkspaceService implements Disposable, IWorkspaceService {
     return ws;
   }
 
+  /** Given a standard vault, convert it into a self contained vault.
+   *
+   * The function **mutates** (modifies) the vault object. */
+  static standardToSelfContainedVault(vault: DVault): SelfContainedVault {
+    if (VaultUtils.isSelfContained(vault)) return vault;
+
+    if (vault.remote?.url) {
+      // Remote vault, calculate path based on the remote
+      vault.fsPath = path.join(
+        FOLDERS.DEPENDENCIES,
+        GitUtils.remoteUrlToDependencyPath({
+          vaultName: vault.name || basename(vault.fsPath),
+          url: vault.remote.url,
+        })
+      );
+    } else {
+      // Local vault, calculate path for local deps
+      vault.fsPath = path.join(
+        FOLDERS.DEPENDENCIES,
+        FOLDERS.LOCAL_DEPENDENCY,
+        path.basename(vault.fsPath)
+      );
+    }
+
+    vault.selfContained = true;
+    // Cast required, because TypeScript doesn't recognize `selfContained` is always set to true
+    return vault as SelfContainedVault;
+  }
+
+  /** Creates a new workspace where the workspace is a self contained vault.
+   *
+   * If the vaults passed to this function are not self contained vaults, they
+   * will be converted to self contained vaults before being created. The vault
+   * objects passed in are **mutated**.
+   *
+   * Further, the first vault given will be the self contained vault that is
+   * used as the workspace root.
+   */
+  static async createSelfContainedVaultWorkspace(
+    opts: WorkspaceServiceCreateOpts
+  ) {
+    const { wsRoot, vaults } = opts;
+    const ws = new WorkspaceService({ wsRoot });
+    if (vaults && vaults.length > 0) {
+      // First vault is the self contained vault we are using as the workspace
+      const wsVault = vaults[0];
+      // The vault is the workspace too
+      if (wsVault.name === undefined) {
+        wsVault.name = path.basename(wsRoot);
+      }
+      wsVault.fsPath = ".";
+      wsVault.selfContained = true;
+
+      // Mutate vault objects to convert them to self contained vaults. The
+      // first vault will be skipped because the conversion is a no-op for
+      // vaults that are already self contained.
+      const selfContainedVaults = vaults.map(
+        WorkspaceService.standardToSelfContainedVault
+      );
+      for (const vault of selfContainedVaults) {
```

Comment:
can we use [[asyncLoopOneAtATime|dendron://dendron.docs/pkg.common-server#asynclooponeatatime]] to avoid the `eslint-disable`?